packages/dds/tree/src/shared-tree/transactionTypes.ts

```diff
@@ -49,7 +49,18 @@ export type TransactionCallbackStatus<TSuccessValue, TFailureValue> =
 			rollback: true;
 			/** The user defined value when the transaction failed. */
 			value: TFailureValue;
-	  };
+	  }
+) & {
+	/**
+	 * An optional list of {@link TransactionConstraint | constraints} that will be checked when the commit corresponding
+	 * to this transaction is reverted.
+	 * If any of the constraints are not met after the transaction callback runs, an error will be thrown. Basically,
+	 * these constraints have to be met after the transaction is applied.
+	 * If any of the constraints are not met when the revert is being applied either locally or on remote clients, the
+	 * revert will be ignored.
```

Comment:
I have a couple of tweaks to suggest:
* I was concerned that alternating between talking about the effects of the constraints at the time the commit is reverted (sentences 1 and 4) and talking about the effects of these constraints on the current transaction (sentences 2 and 3) might be confusing so I think it would be good to move sentence 2 to the end.
* "after the transaction is applied" could be interpreted as including when a peer applies it, which would give the wrong impression. I suggest "at the time they are first introduced".
* Clarifying where the error is thrown from.

```suggestion
	 * An optional list of {@link TransactionConstraint | constraints} that will be checked when the commit corresponding
	 * to this transaction is reverted.
	 * If any of these constraints are not met when the revert is being applied either locally or on remote clients, the
	 * revert will be ignored.
	 * Constraints must be met at the time they are first introduced.
	 * If any of the constraints are not met after the transaction callback returns,
	 * then `runTransaction` (which invokes the transaction callback) will throw a `UsageError`.
```

Reply:
Good suggestions. I updated the comments and separated the parts about validations when running it the first time and when applying the revert.